src/transformers/tokenization_utils_base.py

```diff
@@ -839,7 +839,7 @@ def sanitize_special_tokens(self) -> int:
         """
         return self.add_tokens(self.all_special_tokens_extended, special_tokens=True)
 
-    def add_special_tokens(self, special_tokens_dict: Dict[str, Union[str, AddedToken]]) -> int:
+    def add_special_tokens(self, special_tokens_dict: Dict[str, Union[str, AddedToken]], replace_additional_special_tokens=True) -> int:
```

Comment:
While the name of this method implies calling this method multiple times will update the set of special tokens, it is currently replace the whole set of `self._additional_speical_tokens`.

- with `replace_additional_special_tokens` defaults to `True`:
  - backward compatibility
  - but the general fix desired in PR #20401 will fail T5 tokenizer tests, and maybe affect the loading.
    - We will need to figure out a solution, which is likely be very ugly. 
- with `replace_additional_special_tokens` defaults to `False`:
  - T5 tokenizer test will be OK with respect to a generic fix in PR #20401
  - A few tests has to be fixed (NLLB, etc.)

Reply:
@sgugger @SaulLu Could you take a look and let me know which way we should take. I guess it's 99.99% for `replace_additional_special_tokens` defaults to `True`, but just in case you want to change your mind.

Reply:
@SaulLu 

In general, we want to be able to save the tokenizers with the updated attributes. (I know, your concern is real regarding if we will get the identical tokenizers).

While trying a generic fix (i.e. save the updated attributes for all), there is a problem in T5 Tokenizer:
- `T5Tokenizer` originally have `extra_id_x` (`x` from `1` to `99`) in `additional_special_tokens` - but the initial argument in `__init__` is `None`. These special tokens is  created from `extra_ids=100` in the `__init__`.
- With a generic fix, this `additional_special_tokens` is saved. Previously, it is the one in `init_kwargs` being saved, which was `None`.
- Now when we add a new special token (say `new_token`), previously the whole list is being replaced, but it is still the version in `init_kwargs` being saved. With a generic fix, it is `self.additional_special_tokens = ["new_token"]` being saved.
- Now when we reload the saved tokenizer, it will fail at
https://github.com/huggingface/transformers/blob/afce73bd9d891b55dcb8d4d875d17718ffa01ff0/src/transformers/models/t5/tokenization_t5.py#L132
as it requires `additional_special_tokens` to include all `extra_id_x`

Now if we can allow `additional_special_tokens` being updated instead of being replaced, the saved `additional_special_tokens` will contain `new_token` as well as `extra_id_x`, so the reloading will work.

Reply:
Thank you very much for explaining so clearly the concern you wish to address.

I think this case illustrates that it is the feature that is complicated (but certainly not impossible) to make equivalent 1) a sequence of actions on a tokenizer and 2) a single loading of a tokenizer.

I understand that if the token had been added with the `replace_additional_special_tokens defaults=False` argument it would solve the problem. However it is still a real problem because we don't prevent a user to call the method with `replace_additional_special_tokens defaults=True`  and this user won't be able to reload his tokenizer afterwards. I think we should also have a fix for this case.

Reply:
I agree @SaulLu . However this should be addressed in another PR. As the current PR keeps the current behavior (with the default value of `replace_additional_special_tokens` being `True`), and the generic fix for #20401 is not implemented (so nothing to be fixed yet at this moment).

Again, this PR only tries to
 - make things clear (despite we don't change the method name)
 - provide a possibility to fix more things in a cleaner way in the future
   - It's not 100% certain if we absolutely need the change in this PR for the T5Tokenizer issue w.r.t generic fix mentioned above 
   - **So when reviewing this PR, we should focus on**:
     - if this PR could reduce the confusion of this method  `add_special_tokens`
     - if the change increases the complexity and we don't want to risk

Reply:
Oh yes, I agree that we should not aim to solve the T5Tokenizer ("general fix") issue problem in this PR! 

I had just continued the discussion after the part of your post discussing the pros and cons of the default value to be associated with the `replace_additional_special_tokens` argument. :relaxed: 

Another proposal to clarify what this method does without adding options to it could be to add a note in its docstring.  For example:
```python
def add_special_tokens(self, special_tokens_dict: Dict[str, Union[str, AddedToken]]) -> int:
        """
        Add a dictionary of special tokens (eos, pad, cls, etc.) to the encoder and link them to class attributes. If
        special tokens are NOT in the vocabulary, they are added to it (indexed starting from the last index of the
        current vocabulary).

        Note 1: When adding new tokens to the vocabulary, you should make sure to also resize the token embedding
        matrix of the model so that its embedding matrix matches the tokenizer.

        In order to do that, please use the [`~PreTrainedModel.resize_token_embeddings`] method.

        Using `add_special_tokens` will ensure your special tokens can be used in several ways:

        - Special tokens are carefully handled by the tokenizer (they are never split).
        - You can easily refer to special tokens using tokenizer class attributes like `tokenizer.cls_token`. This
          makes it easy to develop model-agnostic training and fine-tuning scripts.

        When possible, special tokens are already registered for provided pretrained models (for instance
        [`BertTokenizer`] `cls_token` is already registered to be :obj*'[CLS]'* and XLM's one is also registered to be
        `'</s>'`).

        Note 2: If a token or list of tokens was already registered for one of the special attributes mentioned in the
        `special_tokens_dict` argument, it is replaced by the new value in the dictionary. The old tokens are kept in
        the vocabulary. In the particular case of `additional_special_tokens`, if you want to keep the tokens that were
        already present in this list and only add a new token `"<new_token_1>"`, you will have to do

        ```python
        tokenizer.add_special_tokens(
            {
                "additional_special_tokens": [
                    *tokenizer.special_tokens_map_extended["additional_special_tokens"],
                    "<new_token_1>",
                ]
            }
        )
        ```
        """
```

But personally I'm not sure where the balance is between "if this PR could reduce the confusion of the meaning of this method add_special_tokens" and "if the change increases the complexity that we don't want to risk/maintain". I would love the advice of one of the core maintainer on this.

Reply:
One  thing I would like to point out that (again), currently T5 tokenizer will flag all `extra_id_x` as non-special when a user add their own new special tokens through `add_special_tokens`. 

Although I agree your comment in general (i.e. it is ok to replace the original set of special tokens if the user want to), I feel `replace` instead of `update` for **this particular case (T5)** will be very un-intuitive and users are going to complain (if they ever call this method) - why would those originally `extra_id_x` (which is created by default, not by the user) would become non-special? Of course, they can always collect the current special tokens and add new one, and sent the new set as argument to this method.

Let's wait @sgugger and @LysandreJik 's comments though :-)

Reply:
> I feel replace instead of update for this particular case (T5) will be very un-intuitive and users are going to complain (if they ever call this method) - why would those originally extra_id_x (which is created by default, not by the user) would become non-special? Of course, they can always collect the current special tokens and add new one, and sent the new set as argument to this method, but I guess you can imagine this is quite friction.

I can only agree with you on this point! With a part of the research team we are actually twikking a T5 model right now and this behavior was not intuitive. Personally, I would have liked these extra tokens to have their own attribute like the mask token for example in the `SpecialTokensMixin` because these tokens have a role for T5 as important as the mask token for Bert. Btw, I had shared this opinion not so long ago in https://github.com/huggingface/transformers/issues/19298#issuecomment-1321754482.